k2/csrc/timer.h

```diff
@@ -12,37 +12,64 @@
 #ifndef K2_CSRC_TIMER_H_
 #define K2_CSRC_TIMER_H_
 
+#include <mutex>  // NOLINT
+
 #include "k2/csrc/log.h"
 
 namespace k2 {
+
 class Timer {
  public:
   Timer() {
-    K2_CUDA_SAFE_CALL(cudaEventCreate(&time_start_));
-    K2_CUDA_SAFE_CALL(cudaEventCreate(&time_end_));
-    Reset();
+    static std::once_flag has_cuda_init_flag;
```

Comment:
Why doesn't class Timer work for CPU? 
It might make sense for it to take a cudaStream_t, which can be kCudaStreamInvalid if we want it to work on CPU.

Reply:
`cudaEventCreate` returns an error status and `K2_CUDA_SAFE_CALL` will fail.

Reply:
Yeah but what I meant is, there's no reason why that class should be for CPU only. 
It should probably take the Context and behave appropriately.
But I guess these changes are OK as a temporary thing, if we plan to fix it later.

Reply:
> It might make sense for it to take a cudaStream_t, which can be kCudaStreamInvalid if we want it to work on CPU.

How is the `kCudaStreamInvalid` argument going to be used by cpu code?
The previous Cuda code did not take any arguments.

Reply:
> Why doesn't class Timer work for CPU?

Back to the time we created time.h, we actually have not stable context stuff and we only use it to test GPU matrix transpose (it's actually an experiment code that time), that's why we don't write CPU version that time.

Maybe we can just copy code from Kaldi for CPU version https://github.com/kaldi-asr/kaldi/blob/master/src/base/timer.h

Reply:
thanks, I am porting it.